mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,271 @@ def load_model(model_uri):
+           artifact file.
+
+    Note: this class is not thread-safe.
+    Design rule for this class:
+     Because this class instance is a global instance, in order to prevent memory leak, it should
+     only holds IDs and other small objects references. This class internal data structure should
+     avoid reference to user dataset variables or model variables.
+    """
+    def __init__(self):
+        self._pred_result_id_to_dataset_name_and_run_id = {}
+        self._eval_dataset_info_map = defaultdict(lambda: defaultdict(list))
+        self._metric_api_call_info = defaultdict(lambda: defaultdict(list))
+        self._log_post_training_metrics_enabled = True
+        self._metric_info_artifact_need_update = defaultdict(lambda: False)
+
+    def should_log_post_training_metrics(self):
+        """
+        Check whether we should run patching code for autologging post training metrics.
+        This checking should surround the whole patched code due to the safe guard checking,
+        See following note.
+
+        Note: It includes checking `_SklearnTrainingSession.is_active()`, This is a safe guarding
+        for meta-estimator (e.g. GridSearchCV) case:
+          running GridSearchCV.fit, the nested `estimator.fit` will be called in parallel,
+          but, the _autolog_training_status is a global status without thread-safe lock protecting.
+          This safe guarding will prevent code run into this case.
+        """
+        return _SklearnTrainingSession.is_active() and self._log_post_training_metrics_enabled
+
+    def disable_log_post_training_metrics(self):
+
+        class LogPostTrainingMetricsDisabledScope:
+            def __enter__(inner_self):
+                inner_self.old_status = self._log_post_training_metrics_enabled
+                self._log_post_training_metrics_enabled = False
+
+            def __exit__(inner_self, exc_type, exc_val, exc_tb):
+                self._log_post_training_metrics_enabled = inner_self.old_status
+
+        return LogPostTrainingMetricsDisabledScope()
+
+    @staticmethod
+    def get_run_id_for_model(model):
+        return getattr(model, '_mlflow_run_id', None)
+
+    @staticmethod
+    def is_metrics_value_loggable(metric_value):
+        return np.isscalar(metric_value)
+
+    def register_model(self, model, run_id):
+        """
+        In `patched_fit`, we need register the model with the run_id used in `patched_fit`
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_run_id = run_id
+
+    @staticmethod
+    def gen_name_with_index(name, index):
+        assert index >= 0
+        if index == 0:
+            return name
+        else:
+            # Use '-' as the separator between name and index,
+            # The '-' is not valid character in python var name
+            # so it can prevent name conflicts after appending index.
+            return f'{name}-{index + 1}'
```

Comment:
A notable trick:
![image](https://user-images.githubusercontent.com/19235986/126308017-c71da45f-e849-43db-b911-feda954b0749.png)

Reply:
What about using a more special separator like "$" in key name, e.g.
`accuracy_score$5_test_iris_X$2`

Reply:
I think that `-` is just fine! Nice work @WeichenXu123 !